Create equal range entry for cpp (#6829)
Codecademy/docs: content/cpp/concepts/maps/terms/equal-range/equal-range.md

---

> Title: '.equal_range()'
> Description: 'Finds the range of elements equivalent to a specified value in a set or multiset.'
> Subjects:
>   - 'C++'
>   - 'C++ STL'
> Tags:
>   - 'C++'
>   - 'C++ STL'
>   - 'C++ Set'
>   - 'C++ Multiset'
>   - 'C++ Equal Range'
> CatalogContent:
>   - 'C++ STL'
>   - 'C++ Set and Multiset'
>   - 'C++ Algorithms'
>   - 'C++ Standard Library'

Reviewer: ```suggestion
Title: '.equal_range()'
Description: 'Returns a pair of iterators defining the range of elements with the given key'
Subjects:
  - 'Code Foundations'
  - 'Computer Science'
Tags:
  - 'Iterators'
  - 'Map'
  - 'STL'
CatalogContent:
  - 'learn-c-plus-plus'
  - 'paths/computer-science'
```

---

> In C++, the **`.equal_range()`** function is a member function of the `std::set` and `std::multiset` containers. It is used to find the range of elements that are equivalent to a specified value in the set. The function returns a pair of iterators that represent the beginning and end of the range of elements that are equal to the specified value.

Reviewer: ```suggestion
In C++, the **`.equal_range()`** function returns a pair of iterators that define the range of elements matching a given key in an associative container. It is especially useful with `multimap`, where multiple elements can have the same key, allowing you to retrieve all of them efficiently. In a `map`, where keys are unique, the returned range will contain at most one element.
```

### Syntax

```pseudo

> std::pair<iterator, iterator> equal_range(const Key& key);

Reviewer: ```suggestion
mapName.equal_range(key);
```

```

> ### Parameters
> 
> - **`key`**: The value for which the range is to be found.
> - **`iterator`**: The type of iterator used in the set or multiset.
> - **`pair`**: A pair of iterators representing the range of elements equivalent to the specified value.

Reviewer: ```suggestion
**Parameters:**

- `key`: The key value to search for in the map.
```

> ### Return Value
> 
> The function returns a pair of iterators. The first iterator points to the first element in the range that is equivalent to the specified value, and the second iterator points to the first element that is greater than the specified value (i.e., the end of the range).

Reviewer: ```suggestion
**Return value:**
- Returns a `std::pair` of iterators:

  - The first iterator points to the first element equal to the key (or the position where it would be inserted).
  - The second iterator points to the element just past the last element equal to the key.
```

### Example

```cpp
#include <iostream>
#include <set>

int main() {
    std::set<int> mySet = {1, 2, 3, 4, 5};

    // Find the range of elements equal to 3
    auto range = mySet.equal_range(3);

    std::cout << "Elements equal to 3: ";
    for (auto it = range.first; it != range.second; ++it) {
        std::cout << *it << " ";
    }
    std::cout << std::endl;

    return 0;

> }

Reviewer: this should be changed, we need to implement `equal_range()` in map. Here is an example:

```cpp
#include <iostream>
#include <map>

int main() {
  std::map<int, std::string> myMap = {
    {1, "apple"},
    {2, "banana"},
    {3, "cherry"},
    {4, "date"},
    {5, "elderberry"}
  };

  // Find the range of elements with key 3
  auto range = myMap.equal_range(3);

  if (range.first != range.second) {
    std::cout << "Elements with key 3:" << std::endl;
      for (auto it = range.first; it != range.second; ++it) {
        std::cout << it->first << " => " << it->second << std::endl;
      }
  } else {
      std::cout << "No elements with key 3 found." << std::endl;
  }

  return 0;
}
```

```

Output:

```shell

> Elements equal to 3: 3 

Reviewer: change the output accordingly

```

> 

Reviewer: please also add a Codebyte Example section
